### ISSUE TITLE
Old maps are upgraded to ZSTD block compression, making it impossible to load the world in 5.4

### ISSUE THREAD

**Issue by erlehmann** [minetest/minetest#12005]
##### Minetest version
commit 48e508052ad477eda1142f08990d523c8b9701ea

##### Summary
The new, more efficient ZSTD compression is applied to existing worlds.

This makes it impossible to downgrade Minetest from 5.5-dev to 5.4.

##### Steps to reproduce
1. Open a MIneClone 5 world in 5.4.
2. Open the same MIneClone 5 world in 5.5-dev. Play around a bit.
3. Decide that MIneClone 5 does not work yet with 5.5-dev., due to https://github.com/minetest/minetest/issues/11997 or something.
4. Try to open the existing MineClone 5 world in Minetest 5.4 to play some more there.
5. Verify that you can not open the world in Minetest 5.4 anymore.

##### Bug bounties

1. I offer to send 25€ via SEPA to wherever @celeron55 or @rubenwardy tell me it would contribute to Minetest or ContentDB hosting if a setting is included in 5.5 that makes *existing* local maps (including map downloads) not upgrade the map format to ZSTD. Note that I still want new maps to use the better compression, even if the setting is active.

2. I offer to send 25€ via SEPA to wherever @celeron55 or @rubenwardy tell me it would contribute to Minetest or ContentDB hosting if that setting defaults to true in Minetest 5.5.

**Comment by lhofhansl** [comment]
Hmm... I suppose the world could be written as of the version it was created. Would you then make upgrading a manual step? Not sure what that would entail. And I do not need any money. :)

Note that we're talking about the *server* here. In your case you are running the world with a newer server after which you back to an old server. Is that something we generally want to start supporting?

Old/New *clients* are unaffected and backward and forward compatible.

**Comment by Li0nsDickachu** [comment]
> Old/New _clients_ are unaffected and backward and forward compatible.

I am pretty sure this is not true, since players have local single player worlds that would break if you upgrade and then downgrade again (for example due to mod incompatibilities)

**Comment by erlehmann** [comment]
>  And I do not need any money. :)

@lhofhansl the money is offered explicitly for Minetest or ContentDB hosting costs. I do not want to pay developers money directly for fixiing bugs, so that no one feels they should do because they want to make a quick 50 bucks.

To be clear: If *anyone* fixes this, I give money not to the person who fixed it, but to Minetest as a project.

**Comment by lhofhansl** [comment]
> I am pretty sure this is not true, since players have local single player worlds that would break if you upgrade and then downgrade again (for example due to mod incompatibilities)

The singleplayer work starts a server.

**Comment by erlehmann** [comment]
> In your case you are running the world with a newer server after which you back to an old server. Is that something we generally want to start supporting?

It seemed to have worked in the past, or not?

> Singleplayer starts a server.

Yes, and that is where the problem is: Consider some kid upgrades to Minetest 5.5 and will want to go back to Minetest 5.4 once they realize that MineClone5 does not work. Only that they can not do it.

**Comment by lhofhansl** [comment]
I know whining does not help... But. I spent a *lot* of time on adding zstd to Minetest, all in the interest of improving things.
Recently it seems anything added to Minetest spawns endless discussions or whining (even things as innocuous as shadows).

Personally I have been much less motivated to work on the Minetest engine.

**Comment by sfan5** [comment]
> It seemed to have worked in the past, or not?

No, it has not. Upgrades in the map format have always been irreversible. So have node "upgrades" by means of aliases, LBMs or other automatic migrations.
Backups are the only sure way to downgrade.

**Comment by sorcerykid** [comment]
I know that talamh alread has a map reversion script in the works that handles just the use case described here. And if I get time, I might also include a utility as part of my RocketLib Toolkit to convert maps to legacy formats.

It certainly seems that if the precedent has always been that map format upgrades are irreversible, as @sfan5 points out, then that is the most sensible way to move forward. I'm curious, however, if any warning is presented to the end user, so they can make an informed decision before any such conversion even takes place?

**Comment by lhofhansl** [comment]
I'd love a warning when the server is making irreversible changes to a map, and then an option not to start the server - especially for singleplayer.

**Comment by sfan5** [comment]
Note that starting up a map at any time can make (unintended) irreversible changes to it, consider LBMs and node aliases.
The engine obviously can't warn about that but how do you want to make it understandable to users that the warning only covers *certain* changes?

**Comment by sorcerykid** [comment]
I think changes to the internal map block format by the engine are not analogous to changes induced by an LBM or node aliases, because the latter two do not completely break compatibility. The database can still be opened and "repaired" within a previous Minetest version, with some effort. Add to the fact, most LBMs and node aliases are defined by mods, and therefore I believe it is the responsibility of those mod authors to warn of such irreversible alterations as well.

**Comment by sfan5** [comment]
That sounds reasonable.

**Comment by Desour** [comment]
> I believe it is the responsibility of those mod authors to warn of such irreversible alterations as well.

Sadly mods have no way to do this conveniently in-game before the game is started or any mods are loaded, or at least *while* loading.

Maybe it would be good to allow scripts to be executed before mod-loading happens, that can prompt for user input and have access to modstorage and similar.
Or at least add an api for prompting for user input at mod loading, blocking the loading until the user answered.
Ideally such prompts should be done independently of whether the server is started via the terminal or with the gui.

**Comment by sorcerykid** [comment]
@Desour, in the case of mods I'm not suggesting a prompt at runtime. I think a warning published somewhere (whether that be the mod's forum topic, the mod's ContentDB page, or even in the mod's documentation), is plenty sufficient.

For the engine, however, I think compatibility breaking changes should be brought to the attention of the end-user at runtime. Of course, making special note on the Minetest forums is still a good practice too, as sfan5 did [here](https://forum.minetest.net/viewtopic.php?f=18&p=406408#p406408).
